### PR TITLE
refactor: reconstruct option management

### PR DESCRIPTION
**Pull request by wxharry**
## Brief Information

This pull request is in the type of ([more info](https://github.com/angular/angular/blob/main/CONTRIBUTING.md#type) about types):

- [ ] build
- [ ] ci
- [ ] docs
- [ ] feat
- [ ] fix
- [ ] perf
- [x] refactor
- [ ] test

Related issues ([all available keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

- #474 

## Details
- [x] use val-loader to get imported features from [here](https://github.com/wxharry/hypertrons-crx/blob/impl-options/src/pages/ContentScripts/index.ts)
- [x] refactor options management
    - [x] replace class with functions
    - [x] add all imported features to the option list
    - [x] control on/off from the `feature-manager`
- [ ] ~use `webext-options-sync-per-domain` to improve the option management (?)~
- [ ] ~need consistent feature ID, especially for locale~
- [ ] ~cache related~
- [ ] ~docs and comments~

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/hypertrons/hypertrons-crx/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/hypertrons/hypertrons-crx)

### CODE REVIEW

**Comment by wxharry** [comment]
> Nice work! It works as soon as I cancel the option button，but there is a small problem that all the options are re-selected after refreshing.
> 

Thank you for pointing that out. Since this pr is still WIP, i will fix it in the next few commits.

**Comment by tyn1998** [comment]
Hi @wxharry, I have learnt your code especially the "importedFeatures" part and I totally figured out the way you extracted the imported features from `index.ts` and injected them into a webpack module during build. 

I learnt a lot about [webpack loaders](https://juejin.cn/post/6992754161221632030#heading-2) as well, finding that `val-loader` is not so necessary and `features-loader.js` should not be treated as a webpack module, rather it should be used as a webpack loader as its name suggests.

So I changed the code a bit, treating README.md as a module (this is how rgh does, though for now we don't extract feature list from README.md), and used `features-loader.cjs` to convert the README.md module so its content will be `export const importedFeatures = ...` after being processed by the loader. Then we can `import {importedFeatures} from 'path/of/README.md'` anywhere in code.

**Comment by tyn1998** [comment]
There is still a lot of room for code improvement, working in progress...

File: src/utils/settings.ts [hunk 1]
```diff
@@ -9,41 +9,40 @@ export interface FeatureOption {
...
 }
 
+export const defaultSettings: Settings = {
+  featureOptions: [],
+  locale: 'en',
+};
+
 export async function loadSettings() {
-  let settingsConfig = await chromeGet('settings');
+  let settingsConfig = JSON.parse(await chromeGet('settings'));
```
[the remark this answers is not in the document]
Comment (tyn1998) [reply src/utils/settings.ts]: image

Well after I pulled the latest two commits in, the options page was broken in my side. Looking for the cause...

**Comment by tyn1998** [comment]
Actually, what I did last week (those 4 commits) didn't work right because I found several checked checkbox would be unchecked after page refresh this morning.

However hours work later, I hope the option manager now works. I did:

- [x] refactored a more cohesive `options-storage.ts` (a single instance and be consumed everywhere, this time I vote for Class rather several functions)
- [x] to use multiple keys to store all options, rather a single "settings" key whose value is a serialized stirng
- [x] use `chrome.storage.sync.set/get` so options can be synced by google account. (we can pass [default option[s](](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/StorageArea/get#parameters)url) via `get(defaults)`)
- [x] other changes

Hi @lhbvvvvv, could you help to test this PR again?

**Comment by tyn1998** [comment]
@wxharry Thank you for your test! I know what causes the bug and will fix it ASAP.

> Also, I think `perceptor-layout`, `perceptor-tab`, and `repo-networks` can be combined into one option.

Considered this, but maybe improve it later :) thanks for your suggestion!

**Comment by tyn1998** [comment]
@lhbvvvvv fixed it, please verify if my patch works.

Hi @wxharry, could you give a review on my code?

**Comment by wxharry** [comment]
> Hi @wxharry, could you give a review on my code?
LGTM

**Comment by lhbvvvvv** [comment]
> @lhbvvvvv fixed it, please verify if my patch works.

It ran successfully without problems.

**Pull request merged**